Add support for protobuf TxGenerator and SIGN_MODE_DIRECT

aaronc opened the pull request:
ref: #6213 

# Summary

This PR adds support for a protobuf `client.TxGenerator` with signing via `SIGN_MODE_DIRECT`.

# Details

* Protobuf `Tx` additions/changes:
  * Add implementations of `client.TxGenerator` and `client.TxBuilder` for the new protobuf `Tx` in the new `x/auth/tx` package (the old `StdTx` will move to `x/auth/tx/legacy` in a future PR)
  * Add a `SignModeHandler` implementation for `SIGN_MODE_DIRECT`
  * Adds the `TxRaw` proto type and updates `SignDoc` as well as doc comments to reflect the latest state of ADR 020, updates `SignerInfo` to use `PublicKey` rather than `Any`.
* Refactoring of existing code:
  * Move `sdk.TestMsg` to `codec/testdata` adding protobuf support
  * Move the `x/auth/ante.SigVerifiableTx` to `x/auth/signing.SigVerifiableTx` so that it can be used from other packages
  * Add the `SigFeeMemoTx` interface to `x/auth/signing` which wraps `sdk.Tx`, `FeeTx`, `TxWithMemo` and `SigVerifiableTx` in a single interface and use this in `client.TxBuilder`. (I'm not sure this is well named... but functionally it makes sense because this is what `TxBuilder` actually supports and is standard throughout the SDK.)
  * Rename `CompactBitArray.Size()` to `Count()` and re-enable `gogoproto.sizer` - this was causing a nasty bug and the proto `Size` method shouldn't have been overridden.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

alexanderbez commented on client/tx_generator.go:
```diff
@@ -25,7 +25,7 @@ type (
 	// also know how to encode itself.
 	TxBuilder interface {
-		GetTx() sdk.Tx
+		GetTx() signing.SigFeeMemoTx
```
Can we think of a better name for this? What exactly is the intention of this interface?

aaronc replied:
It’s what is actually supported by both the proto buf and amino tx. Maybe it should simply be `Tx` or `StdTx` but that would be confusing for other reasons. Maybe `FullTx`? I’m pretty stumped on names

aaronc replied:
Any ideas? Otherwise maybe we change this in another PR?

aaronc replied:
Added a TODO item to #6213 for this.

fedekunze commented on x/auth/types/test_utils.go:
```diff
@@ -4,11 +4,13 @@ import (
 
-func NewTestMsg(addrs ...sdk.AccAddress) *sdk.TestMsg {
-	return sdk.NewTestMsg(addrs...)
+func NewTestMsg(addrs ...sdk.AccAddress) *testdata.TestMsg {
+	return testdata.NewTestMsg(addrs...)
 }
```
I don't see the need for this func

aaronc replied:
Yeah, it's pointless basically. It's a holdover from before, but will remove.

fedekunze commented on x/auth/tx/sigs.go:
```diff
@@ -0,0 +1,104 @@
+			},
+		}, sig
+	default:
+		panic("unexpected case")
```
can we wrap the panic with the data error? 
```suggestion
		panic(fmt.Sprintf("unexpected signature data type %T", data))
```

fedekunze commented on x/auth/tx/sigs.go:
```diff
@@ -0,0 +1,104 @@
+		}, nil
+
+	default:
+		panic("unexpected case")
```
ditto

fedekunze commented on x/auth/tx/sigs.go:
can we add a test case for this line?

aaronc replied:
added

fedekunze commented on x/auth/tx/encoder.go:
```diff
@@ -0,0 +1,39 @@
+	return func(tx types.Tx) ([]byte, error) {
+		wrapper, ok := tx.(*builder)
+		if !ok {
+			return nil, fmt.Errorf("expected %T, got %T", builder{}, tx)
```
```suggestion
			return nil, fmt.Errorf("expected %T, got %T", &builder{}, tx)
```

fedekunze commented on x/auth/tx/encoder.go:
```diff
@@ -0,0 +1,39 @@
+	return func(tx types.Tx) ([]byte, error) {
+		wrapper, ok := tx.(*builder)
+		if !ok {
+			return nil, fmt.Errorf("expected %T, got %T", builder{}, tx)
```
```suggestion
			return nil, fmt.Errorf("expected %T, got %T", &builder{}, tx)
```

fedekunze commented on x/auth/tx/builder.go:
```diff
@@ -0,0 +1,317 @@
+
+func (t *builder) ValidateBasic() error {
+	theTx := t.tx
+	if theTx == nil {
+		return fmt.Errorf("bad Tx")
+	}
+
+	body := t.tx.Body
+	if body == nil {
+		return fmt.Errorf("missing TxBody")
+	}
+
+	authInfo := t.tx.AuthInfo
+	if authInfo == nil {
+		return fmt.Errorf("missing AuthInfo")
+	}
+
+	fee := authInfo.Fee
+	if fee == nil {
+		return fmt.Errorf("missing fee")
+	}
+
+	if fee.GasLimit > MaxGasWanted {
+		return sdkerrors.Wrapf(
+			sdkerrors.ErrInvalidRequest,
+			"invalid gas supplied; %d > %d", fee.GasLimit, MaxGasWanted,
+		)
+	}
+
+	if fee.Amount.IsAnyNegative() {
+		return sdkerrors.Wrapf(
+			sdkerrors.ErrInsufficientFee,
+			"invalid fee provided: %s", fee.Amount,
+		)
+	}
```
these conditions need unit test cases

aaronc replied:
done

blushi commented on x/auth/signing/direct/direct.go:
```diff
@@ -0,0 +1,66 @@
+var _ signing.SignModeHandler = ModeHandler{}
+
+// DefaultMode implements SignModeHandler.DefaultMode
+func (h ModeHandler) DefaultMode() signingtypes.SignMode {
```
could we omit the receiver identifier `h` here too?

sahith-narahari replied:
we need this as ModeHandler has to satisfy SignModeHandler interface

aaronc replied:
This is just about removing `h` here which makes sense.

sahith-narahari commented on x/auth/tx/builder.go:
```diff
@@ -0,0 +1,317 @@
+	t.authInfoBz = nil
+}
+
+func (t *builder) SetFeeAmount(coins sdk.Coins) {
```
@aaronc wouldn't it be better to have SetFee rather than just feeAmount

aaronc replied:
`SetFeeAmount` was requested in a previous PR for `TxBuilder`. I think it's okay but either way is out of the scope of this PR.

aaronc commented:
@fedekunze addressed all your review comments. The test coverage should be better now.

@alexanderbez the only pending thing I see is figuring out this naming thing. What do you think?

Would be good to get this merged soon to keep #6213 moving.
